packages/core/strapi/lib/services/entity-validator/index.js

```diff
@@ -222,10 +223,116 @@ const createValidateEntity =
         entity,
       },
       { isDraft }
-    ).required();
+    )
+      .test('relations-test', 'check that all relations exist', async function (data) {
+        try {
+          await checkRelationsExist(buildRelationsStore(model.uid, data) || {});
+        } catch (e) {
+          return this.createError({
+            path: this.path,
+            message: e.message,
+          });
+        }
+        return true;
+      })
+      .required();
+
     return validateYupSchema(validator, { strict: false, abortEarly: false })(data);
   };
 
+/**
+ * Builds an object containing all the media and relations being associated with an entity
+ * @param {String} uid of the model
+ * @param {Object} data
+ * @param {Object} relationsStore to be updated and returned
+ * @returns
+ */
+const buildRelationsStore = (uid, data, relationsStore = {}) => {
+  const currentModel = strapi.getModel(uid);
+
+  Object.keys(currentModel?.attributes || {}).forEach((attributeName) => {
+    const attribute = currentModel.attributes[attributeName];
+
+    const value = data[attributeName];
+    if (isEmpty(value) || isNil(value)) {
+      return;
+    }
+
+    switch (attribute.type) {
+      case 'relation': {
+        if (!attribute.target) {
+          return;
+        }
+        // If the attribute type is a relation keep track of all
+        // associations being made with relations.
+        const source = Array.isArray(value) ? value : value.connect || value.set || [];
+        const idArray = source.map((v) => ({ id: v.id || v }));
+
+        relationsStore[attribute.target] = relationsStore[attribute.target] || [];
+        relationsStore[attribute.target].push(...idArray);
+        break;
+      }
+      case 'media': {
+        // For media attribute types keep track of all media associated with
+        // this entity.
+        const mediaUID = 'plugin::upload.file';
+        castArray(value).forEach((v) => {
+          relationsStore[mediaUID] = relationsStore[mediaUID] || [];
+          relationsStore[mediaUID].push({ id: v.id || v });
+        });
+        break;
+      }
```

Comment:
Since media are relations, I think we can also expect the connect/set logic here. Maybe you can merge with the code above?

Reply:
Attribute types `media` don't seem to have a target attribute. It seems better to me to treat them separately so we can specify `mediaUID` as the target each time. WDYT?

Reply:
Update: I've addressed this and combined the relation and media cases